crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -133,6 +133,64 @@ where
         let latest_historical = self.database.history_by_block_hash(anchor_hash)?;
         Ok(self.canonical_in_memory_state.state_provider(state.hash(), latest_historical))
     }
+
+    /// Returns the block number, in-block transaction index, and whether the transaction is in
+    /// memory.
+    fn block_number_by_tx_id(
+        &self,
+        id: TxNumber,
+    ) -> ProviderResult<Option<(BlockNumber, usize, bool)>> {
```

Comment:
maybe a enum?

Reply:
When it is in memory, could we just return the entire `BlockState` as well as the tx index here, rather than just the block number? That way the calling methods do not need to fetch the state again

Reply:
yeah right, fixed